core/iwasm/compilation/aot_llvm_extra.cpp

```diff
@@ -382,48 +282,47 @@ aot_func_disable_tce(LLVMValueRef func)
     F->setAttributes(Attrs);
 }
 
-#if LLVM_VERSION_MAJOR >= 12
 void
-aot_apply_llvm_new_pass_manager(AOTCompContext *comp_ctx)
+aot_apply_llvm_new_pass_manager(AOTCompContext *comp_ctx, LLVMModuleRef module)
 {
-    Module *M;
     TargetMachine *TM = unwrap(comp_ctx->target_machine);
-    bool disable_llvm_lto = false;
-
-    LoopAnalysisManager LAM;
-    FunctionAnalysisManager FAM;
-    CGSCCAnalysisManager CGAM;
-    ModuleAnalysisManager MAM;
-
     PipelineTuningOptions PTO;
     PTO.LoopVectorization = true;
     PTO.SLPVectorization = true;
     PTO.LoopUnrolling = true;
 
+#ifndef NDEBUG
+    PassInstrumentationCallbacks PIC;
+    PassBuilder PB(TM, PTO, None, &PIC);
+#else
 #if LLVM_VERSION_MAJOR == 12
     PassBuilder PB(false, TM, PTO);
 #else
     PassBuilder PB(TM, PTO);
+#endif
 #endif
 
+    // Register all the basic analyses with the managers.
+    LoopAnalysisManager LAM;
+    FunctionAnalysisManager FAM;
+    CGSCCAnalysisManager CGAM;
+    ModuleAnalysisManager MAM;
+    PB.registerFunctionAnalyses(FAM);
+    PB.registerLoopAnalyses(LAM);
+    PB.registerModuleAnalyses(MAM);
+    PB.registerCGSCCAnalyses(CGAM);
+    PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
     // Register the target library analysis directly and give it a
     // customized preset TLI.
     std::unique_ptr<TargetLibraryInfoImpl> TLII(
         new TargetLibraryInfoImpl(Triple(TM->getTargetTriple())));
     FAM.registerPass([&] { return TargetLibraryAnalysis(*TLII); });
```

Comment:
Had better not remove L415 to L417, and should not change the order of L409-413 and PB.registerXXX():
https://github.com/WasmEdge/WasmEdge/blob/master/lib/aot/compiler.cpp#L5141-L5150
https://github.com/llvm/llvm-project/blob/main/llvm/lib/LTO/ThinLTOCodeGenerator.cpp#L254-L265

In my test, the CoreMark score of AOT mode is decreased from 26100 to 25100 by this PR, and after restoring the code in aot_emit_function.c and code of this place, the score is restored.

Reply:
The code had better be like:
```C
    LoopAnalysisManager LAM;
    FunctionAnalysisManager FAM;
    CGSCCAnalysisManager CGAM;
    ModuleAnalysisManager MAM;

    // Register the target library analysis directly and give it a
    // customized preset TLI.
    std::unique_ptr<TargetLibraryInfoImpl> TLII(
        new TargetLibraryInfoImpl(Triple(TM->getTargetTriple())));
    FAM.registerPass([&] { return TargetLibraryAnalysis(*TLII); });

    // Register the AA manager first so that our version is the one used.
    AAManager AA = PB.buildDefaultAAPipeline();
    FAM.registerPass([&] { return std::move(AA); });

    // Register all the basic analyses with the managers.
    PB.registerFunctionAnalyses(FAM);
    PB.registerLoopAnalyses(LAM);
    PB.registerModuleAnalyses(MAM);
    PB.registerCGSCCAnalyses(CGAM);
    PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
```